### ISSUE TITLE
Test::Selenium::PhantomJS/Firefox/Chrome?

### ISSUE THREAD

**Issue by amacleay** [gempesaw/Selenium-Remote-Driver#250]
Hello!  This is an inquiry and perhaps doesn't belong in the Issues tracker - sorry if this isn't the right place to ask!

Is there a currently supported way to get an instance of Selenium::PhantomJS (or any of the other packages that consume Selenium::CanStartBinary::FindBinary) to also inherit the behavior of Test::Selenium::Remote::Driver (consumption of Test::Selenium::Remote::Role::DoesTesting together with those args and functions)?

I can see naive ways to achieve this, but what I've imagined isn't terribly satisfactory. I'm wondering if there exists such a mechanism already, and if not and it were helpful, I'd be happy to put together a pull request.

Thanks!

**Comment by gempesaw** [comment]
Oh, hm, I haven't thought about that before. Off the top of my head, we could make a couple classes that `with`s both roles... Would there be any complications? Is that what you had already thought of?

**Comment by amacleay** [comment]
Yeah, I hadn't thought much further than an implementation like this:
```
package Test::Selenium::PhantomJS;
extends 'Selenium::PhantomJS';
sub func_list { ... cp/paste from Test::Selenium::Remote::Driver... }
sub has_args { ... cp/paste... }
with 'Test::Selenium::Remote::Role::DoesTesting';
```
I tested locally with something of the sort, and it seemed fine functionally, but it has the ugly copy/paste.  Extending `Test::Selenium::Remote::Driver` instead has the same problem, but inverted, where code from `Selenium::{PhantomJS,Chrome,Firefox}` has to be copied around.

I suppose a solution to that might be to factor out that particular `has_args` and `func_list` into a new `Test::Selenium::Remote::Driver::DoesTesting` role or something of the sort?

**Comment by gempesaw** [comment]
I think moo can do multiple extends as well, perhaps that's the way to go? *ponder*

**Comment by amacleay** [comment]
I wouldn't argue against it necessarily.  There's of course plenty of discussion about why multiple inheritance is bad, but I suppose a refactor could always be done later were it found to be appropriate.

**Comment by gempesaw** [comment]
```perl
use strict;
use warnings;
use Test::More;

package Test::Selenium::Firefox {
    use Moo;
    extends 'Selenium::Firefox', 'Test::Selenium::Remote::Driver';

    1;
};

my $f = Test::Selenium::Firefox->new;

$f->get_ok('about:config');

done_testing;
```

This very minimal example, and the Test::Selenium::Firefox package is pretty straightforward. Does this seem useful to you?

**Comment by amacleay** [comment]
Yep, it does indeed

**Comment by gempesaw** [comment]
closed by #267 and subsequently #269, thanks!
